openforcefield/topology/molecule.py

```diff
@@ -3176,6 +3341,319 @@ def from_openeye(oemol, allow_undefined_stereo=False):
+        -------
+        molecule : openforcefield.topology.Molecule
+            An openforcefield molecule instance.
+
+        Raises
+        -------
+        AttributeError : if the record dict can not be made from a record instance.
+            if a client is passed, because the client could not retrive the initial molecule.
+
+        KeyError : if the dict does not contain the canonical_isomeric_explicit_hydrogen_mapped_smiles.
+
+        InvalidConformerError : silent error, if the conformer could not be attached.
+        """
+
+        # We can accept the TDeEntry record or the dict with JSON encoding
+        # lets get it all in the dict rep
+        if not isinstance(qca_record, dict):
+            try:
+                qca_record = qca_record.dict(encoding='json')
+            except AttributeError:
+                raise AttributeError('The object passed could not be converted to a dict with json encoding')
+
+        try:
+            mapped_smiles = qca_record['attributes']['canonical_isomeric_explicit_hydrogen_mapped_smiles']
+        except KeyError:
+            raise KeyError('The record must contain the hydrogen mapped smiles to be safley made from the archive.')
+
+        # make a new molecule that has been reordered to match the cmiles mapping
+        offmol = cls.from_mapped_smiles(mapped_smiles, allow_undefined_stereo=allow_undefined_stereo)
+
+        if client is not None:
+            # try and find the initial molecule conformations and attach them
+            # collect the input molecules
+            try:
+                input_mols = client.query_molecules(id=qca_record['initial_molecules'])
+            except KeyError:
+                # this must be an optimisation record
+                input_mols = client.query_molecules(id=qca_record['initial_molecule'])
+            except AttributeError:
+                raise AttributeError('The provided client can not query molecules, make sure it is an instance of'
+                                     'qcportal.client.FractalClient() with the correct address.')
+            initial_ids = {}
+            # now for each molecule convert and attach the input geometry
+            for molecule in input_mols:
+                geometry = unit.Quantity(np.array(molecule.geometry, np.float), unit.bohr)
+                try:
+                    offmol.add_conformer(geometry.in_units_of(unit.angstrom))
+                    initial_ids[molecule.id] = offmol.n_conformers - 1
+                except InvalidConformerError:
+                    print('Invalid conformer for this molecule, the geometry could not be attached.')
+            # attach a dict that has the initial molecule ids and the number of the conformer it is stored in
+            offmol._properties['initial_molecules'] = initial_ids
+
+        return offmol
+
+    @classmethod
+    @RDKitToolkitWrapper.requires_toolkit()
+    def from_pdb_and_smiles(cls, file_path, smiles, allow_undefined_stereo=False):
```

Comment:
@j-wags this is the implementation of the from PDB and smiles classmethod, I think this might be the better option as this way make sure we have the bond orders and stereochemistry assigned when making a molecule from smiles, then we just reorder the atoms and transfer the coordinates from the PDB, we could be even safer and instead reorder the coordinates in the PDB and attach them without interfering with the molecule from smiles.
They only way I could see to do this was to move the code out of a toolkit wrapper however, is this okay?

Reply:
This is a nifty implementation. It's great that it conforms to the PDB ordering in the end.

> They only way I could see to do this was to move the code out of a toolkit wrapper however, is this okay?

It could be OK if there's no alternative, but could you expand on why this couldn't be in RDKitToolkitWrapper. Is it an import error?

Reply:
Thanks and I think that is the most important part here as people will expect the ordering to be the same.

I also think this can be moved to the RDKitToolkitWrapper and have tried this in the latest commit, I think my original plan was letting people read PDB files with the RDKit `from_file` function then trying to catch this and create the molecule but I was worried some molecules directly from the PDB may slip through.

This also opens up the possibility of extending the `add_conformer` API to accept files with the molecules in any order as we can now graph match the structures provided connectivity and reorder the coordinates in the file and attach them to the molecule. This may be best placed in another PR though?

Reply:
> This may be best placed in another PR though?

Agreed. This would be extremely nifty, but we can save it for a later day.